popup.js

```diff
@@ -1,18 +1,119 @@
-    console.log(chrome.tabs);
-    chrome.tabs.query({active: true, currentWindow: true}, function(tabs) {
-        console.log(tabs);
-    });
+function setup() {
+    document.addEventListener('DOMContentLoaded', async function() {
+        // fetch language data from the API and dynamically load elements into the drop-down menus.
+        const resp = await fetch('https://libretranslate.eownerdead.dedyn.io/languages');
+        const languageData =  await resp.json();
+        loadLanguages(languageData);
+
+        // setup local storage for the extension to access language features.
+        /*
+        1. Initialise the user's language preferences during translation.
+        2. Set the selected property of the drop-down menus' options accordingly.
+         */
+        const prefs = await initialiseLanguageSelection();
+        console.log('What is prefs? ' + JSON.stringify(prefs));
+        initialiseLanguagePreferences(prefs.source, prefs.target);
 
-    const downloadButton = document.querySelector('button');
-    if (downloadButton) {
-        downloadButton.addEventListener('click', function() {
-            setupDownloadButton();
+        sourceLanguageListener();
+        targetLanguageListener();
+
+        console.log('dom content loaded');
+        console.log(chrome);
+        console.log(chrome.tabs);
+        chrome.tabs.query({active: true, currentWindow: true}, function(tabs) {
+            console.log(tabs);
         });
+
+        const downloadButton = document.querySelector('button');
+        if (downloadButton) {
+            downloadButton.addEventListener('click', function() {
+                setupDownloadButton();
+            });
+        }
+    });
+}
+
+/*
+Function loadLanguages collects a list of languages and appends them to both
+drop-down menus.
+ */
+function loadLanguages(languages) {
+    const source = document.querySelector('#source-lang');
+    const target = document.querySelector('#target-lang');
+    for (const language of languages) {
+        const sourceOption = document.createElement('option');
+        sourceOption.value = language.code;
+        sourceOption.text = language.name + '.';
+        const targetOption = sourceOption.cloneNode(true);
+
+        source.add(sourceOption);
+        if (targetOption.text !== 'English.') {
+            target.add(targetOption);
+        }
     }
-});
+
+    return source;
+}
+
+/*
+Function initialiseLanguageSelection retrieves the user's language preferences
+for how they translate words.
+The data is used to provide visible feedback in drop-down menus and to set the
+correct payload for querying the translation API.
+ */
+async function initialiseLanguageSelection() {
+    let source = await browser.storage.local.get('source');
+    console.log('Getting source lang on startup: ' + JSON.stringify(source));
+
+    // check if source is an empty object.
+    if (Object.keys(source).length === 0 && source.constructor === Object) {
```

Comment:
there's probably a much cleaner way to do this

Reply:
Thanks for catching this. We can now check the length of the extension's local storage directly to see if it is empty. Also clarified the comments so they describe the if-statements more clearly.

Reply:
Have to leave an update: the new changes weren't working so I went with 
```js
if (!Object.keys(source).length) {
    // code here
}
```
which is the best compromise I could settle with. 

I can't seem to find a solution that behaves like `source.isEmpty()` which would look infinitely better than what I've implemented. I could make my own function that wraps this but I felt that could be discussed first.